Specialize `TrustedLen` for `Iterator::unzip()`

ChayimFriedman2 opened the pull request:
Don't check the capacity every time (and also for `Extend` for tuples, as this is how `unzip()` is implemented).

I did this with an unsafe method on `Extend` that doesn't check for growth (`extend_one_unchecked()`). I've marked it as perma-unstable currently, although we may want to expose it in the future so collections outside of std can benefit from it. Then specialize `Extend for (A, B)` for `TrustedLen` to call it.

An alternative way of implementing this is to have a semi-public trait (`#[doc(hidden)]` public, so collections outside of core can implement it) for `extend()` inside tuples, and specialize it from collections. However:

 1. This looks more complex to me.
 2. This prohibits the option of exposing this somewhen to collections outside of std, as we never expose specializations.

A concern that may arise with the current approach is that implementing `extend_one_unchecked()` correctly must also incur implementing `extend_reserve()`, otherwise you can have UB. This is a somewhat non-local safety invariant. However, I believe this is fine, since to have actual UB you must have unsafe code inside your `extend_one_unchecked()` that makes incorrect assumption, *and* not implement `extend_reserve()`. I've also documented this requirement.

**Benchmark:**

Code:
```rust
#[bench]
fn unzip(b: &mut Bencher) {
    b.iter(|| {
        for _ in 0..10_000 {
            let v: (Vec<_>, VecDeque<_>) =
                (black_box(0u32)..black_box(1_000)).map(|i| (i, i * 2)).unzip();
            black_box(v);
        }
    });
}
```

Before:
```
unzip::unzip 14.17ms/iter  +/- 374.85µs
```
After:
```
unzip::unzip 5.33ms/iter  +/- 164.54µs
```

the8472 commented on library/core/src/iter/traits/collect.rs:
```diff
@@ -427,6 +429,22 @@ pub trait Extend<A> {
+    ///
+    /// This must only be called when we know the collection has enough capacity to contain the new item,
+    /// for example because we previously called `extend_reserve`.
+    #[unstable(feature = "extend_one_unchecked", issue = "none")]
+    #[doc(hidden)]
+    unsafe fn extend_one_unchecked(&mut self, item: A) {
+        self.extend_one(item);
+    }
```
I don't think adding this method to a trait makes sense without intent for stabilization.

Either it should be an internal helper trait or there has to be a discussion whether this should become a public API. Hidden unstable APIs are only done to work around limitations of the type system, not as a "well, we might or might not stabilize it" in-betweens.

Additionally this is only used by a specilization. Since the path to stabilization for specialization is unclear we also refrain from adding APIs related to specializations.

ChayimFriedman2 replied:
@the8472 I was trying to implement what you suggested, but I've hit obstacles: the compiler doesn't accept my code (https://github.com/ChayimFriedman2/rust/tree/nonworking-tuple-extend-spec) because:
```
error: cannot specialize on trait `collect::Extend`
   --> library\core\src\iter\traits\collect.rs:512:14
    |
512 |     ExtendB: Extend<B>,
    |              ^^^^^^^^^

error: cannot specialize on trait `collect::Extend`
   --> library\core\src\iter\traits\collect.rs:511:14
    |
511 |     ExtendA: Extend<A>,
    |              ^^^^^^^^^
```
(A false positive, the compiler is not smart enough to understand this is not a bad specialization, but) And:
```
error: specializing impl repeats parameter `T`
    --> library\alloc\src\collections\vec_deque\mod.rs:2854:1
     |
2854 | impl<T, A: Allocator> ExtendUnchecked<T> for VecDeque<T, A> {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
```
I'm afraid today's specialization cannot express that (can you find a way?). So, considering that, and assuming we'll mark the method as perma-unstable, do you still believe we shouldn't do it?

the8472 replied:
I have not been able to find a better approach either. `min_specialization` is quite limiting. Can you leave a (non-doc) comment on the method that it's an internal method and is only on the public trait due to limitations of specialization?

And add a `where Self: Sized` bound so it won't be added to Extend vtables, they have no use for it.

the8472 reviewed (changes requested):
Mostly looks fine to me and I can't think of a better approach given the current specialization limitations.

Just a few minor points

the8472 commented on library/core/src/iter/traits/collect.rs:
```diff
@@ -466,33 +484,102 @@ where
-        let (lower_bound, _) = iter.size_hint();
+        let (lower_bound, upper_bound) = self.size_hint();
+
+        if upper_bound.is_none() {
+            // We cannot reserve more than `usize::MAX` items, and this is likely to go out of memory anyway.
+            default_extend_tuple(self, a, b);
+            return;
+        }
```
Ah, good point.

Now that makes me wonder how this interacts with HashSets and whether those might accidentally OOM even when it's not guaranteed that those elements will take up extra space.

hashbrown implements it as `reserve`, not `try_reserve`.
https://github.com/rust-lang/hashbrown/blob/4c824c548e85d75f95e0dfcee09c62c5203f9a75/src/map.rs#L6524

Maybe the interface of `extend_reserve` should be changed to return an allocation error so the caller.
But the old implementation also uses `extend_reserve` so it does not have to block this PR.

ChayimFriedman2 committed:
Specialize `TrustedLen` for `Iterator::unzip()`

Don't check the capacity every time (and also for `Extend` for tuples, as this is how `unzip()` is implemented).

I did this with an unsafe method on `Extend` that doesn't check for growth (`extend_one_unchecked()`). I've marked it as perma-unstable currently, although we may want to expose it in the future so collections outside of std can benefit from it. Then specialize `Extend for (A, B)` for `TrustedLen` to call it.

It may seem that an alternative way of implementing this is to have a semi-public trait (`#[doc(hidden)]` public, so collections outside of core can implement it) for `extend()` inside tuples, and specialize it from collections. However, it is impossible due to limitations of `min_specialization`.

A concern that may arise with the current approach is that implementing `extend_one_unchecked()` correctly must also incur implementing `extend_reserve()`, otherwise you can have UB. This is a somewhat non-local safety invariant. However, I believe this is fine, since to have actual UB you must have unsafe code inside your `extend_one_unchecked()` that makes incorrect assumption, *and* not implement `extend_reserve()`. I've also documented this requirement.

ChayimFriedman2 commented on library/core/src/iter/traits/collect.rs:
```diff
@@ -466,33 +484,102 @@ where
+        b.extend_reserve(lower_bound);
+    }
 
+    iter.fold((), extend(a, b));
```
I don't see a reason to unify them. It will take more work and LOC that the few duplicated LOC.

ChayimFriedman2 commented:
@rustbot ready

Did the requested changes.

the8472 commented:
unzip is used in a few places in the compiler, so it might affect perf results

@bors r+ rollup=never

bors commented:
:pushpin: Commit 54556f49d368ba96a92fd6ec352cf6b12e9542ef has been approved by `the8472`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=the8472 54556f49d368ba96a92fd6ec352cf6b12e9542ef -->
<!-- homu: {"type":"Approved","sha":"54556f49d368ba96a92fd6ec352cf6b12e9542ef","approver":"the8472","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 54556f49d368ba96a92fd6ec352cf6b12e9542ef with merge 382148d9a21ae0b506adf44fc1e55a410acde828...
<!-- homu: {"type":"BuildStarted","head_sha":"54556f49d368ba96a92fd6ec352cf6b12e9542ef","merge_sha":"382148d9a21ae0b506adf44fc1e55a410acde828"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9826935840/job/27130890623)
Approved by: the8472
Pushing 382148d9a21ae0b506adf44fc1e55a410acde828 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"the8472","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9826935840/job/27130890623"},"merge_sha":"382148d9a21ae0b506adf44fc1e55a410acde828"} -->

rust-timer commented:
Finished benchmarking commit (382148d9a21ae0b506adf44fc1e55a410acde828): [comparison URL](https://perf.rust-lang.org/compare.html?start=98dcbae5c9ac615d5acfbf42d42b19a77cb9ec11&end=382148d9a21ae0b506adf44fc1e55a410acde828&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=98dcbae5c9ac615d5acfbf42d42b19a77cb9ec11&end=382148d9a21ae0b506adf44fc1e55a410acde828&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=98dcbae5c9ac615d5acfbf42d42b19a77cb9ec11&end=382148d9a21ae0b506adf44fc1e55a410acde828&stat=max-rss)
<details>
<summary>Results (primary 2.0%, secondary 3.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.0% | [1.4%, 2.5%] | 4     |
| Regressions ❌ <br /> (secondary)  | 3.4% | [2.8%, 3.9%] | 3     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.0% | [1.4%, 2.5%] | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=98dcbae5c9ac615d5acfbf42d42b19a77cb9ec11&end=382148d9a21ae0b506adf44fc1e55a410acde828&stat=cycles:u)
<details>
<summary>Results (secondary -3.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.3% | [-3.6%, -3.0%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=98dcbae5c9ac615d5acfbf42d42b19a77cb9ec11&end=382148d9a21ae0b506adf44fc1e55a410acde828&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 699.867s -> 699.903s (0.01%)
**Artifact size**: 328.42 MiB -> 328.38 MiB (-0.01%)
